File: leet_code_algorithms/ransom_note/ransom_note.py
# My Answer

import collections


def canConstruct(self, ransomNote: str, magazine: str) -> bool:
    ransom_note = collections.Counter(ransomNote)
    magazine_note = collections.Counter(magazine)
    for letter in ransom_note:
        if letter in magazine_note:
            if ransom_note[letter] > magazine_note[letter]:
                return False
        else:
            return False
    return True


# Optimal Solution
def canConstruct(self, ransomNote: str, magazine: str) -> bool:
    for i in ransomNote:
        if i in magazine:
            magazine = magazine.replace(i, "", 1)
            # third argument is count:
            # Optional. A number specifying how many occurrences of the old value you want to replace. Default is all occurrences
        else:
            return False

    return True

File: leet_code_algorithms/ransom_note/Ransom_note.java
package ransom_note;

public class Ransom_note {
    // My Solution
    public boolean canConstruct(String ransomNote, String magazine) {
        char[] ransom_array = ransomNote.toCharArray();
        char[] magazine_array = magazine.toCharArray();
        StringBuilder magazine_string_builder = constructStringBuilder(magazine_array);
        for (char letter : ransom_array) {
            if (magazine_string_builder.chars().anyMatch(c -> c == letter)) {
                int index = magazine_string_builder.toString().indexOf(String.valueOf(letter));
                magazine_string_builder = magazine_string_builder.deleteCharAt(index);
            } else {
                return false;
            }
        }
        return true;
    }

    public StringBuilder constructStringBuilder(char[] characters) {
        StringBuilder build = new StringBuilder();
        for (char letter : characters) {
            build.append(letter);
        }
        return build;
    }

    // Optimal Solution
    public boolean canConstructOptimal(String ransomNote, String magazine) {
        if (ransomNote.length() > magazine.length())
            return false;
        int counter[] = new int[26];
        for (char c : magazine.toCharArray()) {
            counter[c - 'a']++;
        }
        for (char c : ransomNote.toCharArray()) {
            if (counter[c - 'a'] == 0)
                return false;
            else
                counter[c - 'a']--;
        }
        return true;
    }
}